### PR TITLE
[Kernels] Add fp8 support to `reshape_and_cache_flash`

### PR DESCRIPTION
**Pull request by Yard1**
Preparation for fp8 support in flash attn-based backends.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
cc @mgoin @tlrmchlsmth

**Comment by robertgshaw2-neuralmagic** [comment]
👀 @Yard1 can you coordinate with @mgoin here? He has been working on separating k_scale and v_scale in the ckpts

**Review by tlrmchlsmth — APPROVED** [review]
lgtm -- looks straightforward, following along with `reshape_and_cache_kernel` is doing.

**Comment by mgoin** [comment]
Hey @Yard1 thanks for sharing - I was working on enabling fp8 for flashinfer in a similar fashion. However I put a pause on the integration to better understand why the performance didn't match expectations. Please check out this [sheet of benchmarks](https://docs.google.com/spreadsheets/d/1qUudqReXoF_HF_rnvCj74bej6qhST4wczMOGHDpv_CI/edit?usp=sharing). I took the benchmarking configuration directly from the FP8 section of their [launch blog](https://flashinfer.ai/2024/02/02/introduce-flashinfer.html). Do you have an idea of why the A100 performance seems so bad in my benchmark?

**Review by comaniac — APPROVED** [review]

**Comment by Yard1** [comment]
@mgoin I hadn't looked into that myself yet. Have you tried asking the flashinfer author in flashinfer repo?

File: vllm/utils.py [hunk 1]
```diff
@@ -507,7 +506,13 @@ def create_kv_caches_with_random_flash(
         key_value_cache = torch.empty(size=key_value_cache_shape,
                                       dtype=torch_dtype,
                                       device=device)
-        key_value_cache.uniform_(-scale, scale)
+        if cache_dtype in ["auto", "half", "bfloat16", "float"]:
+            key_value_cache.uniform_(-scale, scale)
+        elif cache_dtype == 'fp8':
```
Comment (comaniac) [inline vllm/utils.py:511]: Should cover `fp8` and `fp8_e4m3` (they are the same tho).

File: vllm/utils.py [hunk 2]
```diff
@@ -507,7 +506,13 @@ def create_kv_caches_with_random_flash(
         key_value_cache = torch.empty(size=key_value_cache_shape,
                                       dtype=torch_dtype,
                                       device=device)
-        key_value_cache.uniform_(-scale, scale)
+        if cache_dtype in ["auto", "half", "bfloat16", "float"]:
```
Comment (rkooo567) [inline vllm/utils.py:509]: is there no case "auto" becomes fp8?

Comment (rkooo567) [reply vllm/utils.py:509]: besides, handling auto here is a little ugly.. (feel like the caller should convert it alrady) but don't need to handle it here

**Review by rkooo567 — APPROVED** [review]

File: vllm/utils.py (hunk 2 above)
Comment (Yard1) [reply vllm/utils.py:509]: I am just following the pattern laid out in the non-flash function

File: vllm/utils.py (hunk 1 above)
Comment (Yard1) [reply vllm/utils.py:511]: I am just following the pattern laid out in the non-flash function

File: vllm/utils.py (hunk 2 above)
Comment (comaniac) [reply vllm/utils.py:509]: "auto" now means following the model weights/checkpoints. We should definitely improve the logic in the future.

**Pull request merged**